cpp/src/arrow/json/chunker.cc

```diff
@@ -0,0 +1,212 @@
+    ConsumeWhitespace(&partial);
+    if (partial.size() == 0) {
+      // if partial is empty, don't bother looking for completion
+      *completion = string_view();
+      return Status::OK();
+    }
+    auto first_newline = block.find_first_of("\n\r");
+    if (first_newline == string_view::npos) {
+      // no newlines in this block; straddling object straddles *two* block boundaries.
+      // retry with larger buffer
+      return StraddlingTooLarge();
+    }
+    *completion = block.substr(0, first_newline + 1);
+    return Status::OK();
+  }
+};
+
+/// RapidJson custom stream for reading JSON stored in multiple buffers
+/// http://rapidjson.org/md_doc_stream.html#CustomStream
+class MultiStringStream {
+ public:
+  using Ch = char;
+  explicit MultiStringStream(std::vector<string_view> strings)
+      : strings_(std::move(strings)) {
+    std::remove(strings_.begin(), strings_.end(), string_view(""));
+    std::reverse(strings_.begin(), strings_.end());
+  }
+  char Peek() const {
+    if (strings_.size() == 0) return '\0';
+    return strings_.back()[0];
+  }
+  char Take() {
+    if (strings_.size() == 0) return '\0';
+    char taken = strings_.back()[0];
+    if (strings_.back().size() == 1) {
+      strings_.pop_back();
+    } else {
+      strings_.back() = strings_.back().substr(1);
+    }
+    ++index_;
+    return taken;
+  }
+  std::size_t Tell() { return index_; }
+  void Put(char) { ARROW_LOG(FATAL) << "not implemented"; }
+  void Flush() { ARROW_LOG(FATAL) << "not implemented"; }
+  char* PutBegin() {
+    ARROW_LOG(FATAL) << "not implemented";
+    return nullptr;
+  }
+  std::size_t PutEnd(char*) {
+    ARROW_LOG(FATAL) << "not implemented";
+    return 0;
+  }
+
+ private:
+  std::size_t index_ = 0;
+  std::vector<string_view> strings_;
+};
+
+template <typename Stream>
+std::size_t ConsumeWholeObject(Stream&& stream) {
+  static constexpr unsigned parse_flags = rapidjson::kParseIterativeFlag |
+                                          rapidjson::kParseStopWhenDoneFlag |
+                                          rapidjson::kParseNumbersAsStringsFlag;
+  rapidjson::BaseReaderHandler<rapidjson::UTF8<>> handler;
+  rapidjson::Reader reader;
+  // parse a single JSON object
+  switch (reader.Parse<parse_flags>(stream, handler).Code()) {
+    case rapidjson::kParseErrorNone:
+      return stream.Tell();
+    case rapidjson::kParseErrorDocumentEmpty:
+      return 0;
+    default:
+      // rapidjson emitted an error, the most recent object was partial
+      return string_view::npos;
+  }
+}
+
+class ParsingChunker : public Chunker {
+ public:
+  Status Process(string_view block, string_view* chunked) override {
+    if (block.size() == 0) {
+      *chunked = string_view();
+      return Status::OK();
+    }
+    std::size_t total_length = 0;
+    for (auto consumed = block;; consumed = block.substr(total_length)) {
+      auto length = ConsumeWholeObject(rapidjson::StringStream(consumed.data()));
```

Comment:
Passing `StringStream` doesn't sound right, as it expects a zero-terminated string but we don't guarantee that. Instead we should define our own rapidjson stream that explicitly checks for size.

Reply:
Actually it seems the required abstractions are already provided by rapidjson. We could use `EncodedInputStream<UTF8<>, MemoryStream>`.

Reply:
rapidjson considers `'\0'` as the authoritative end of a stream; [`rapidjson::MemoryStream::Peek`](https://github.com/Tencent/rapidjson/blob/7e68aa0a21b7800ec98133cb106e49bd6536e25c/include/rapidjson/memorystream.h#L45) first checks length then returns '\0' for past-the-end locations (then parsers still check whether characters are `== '\0'`).

We can provide the guarantee by padding buffers read from disk with at least one zero byte, which will remove a branching in the parse step.

Reply:
That wouldn't be desirable for e.g. memory-mapped files.

Reply:
Alright, I'll use the explicitly sized stream